skip loading remote data blocks on unsupported WP versions

shekharnwagh opened the pull request:
## Description

Adds version compatibility check before loading the remote data blocks integration. The integration will only load if the current WordPress version meets or exceeds the minimum_wp_version specified in config. If no minimum version is configured, the integration loads normally. This check is required as currently Remote Data Blocks plugin relies on functionality not available in older versions of WordPress (<6.7).

## Changelog Description

<!-- Changelogs are published for our customers. Well-written entries help them stay informed on platform changes and all of the great work that we do! -->

<!-- Write a concise description of changes in the relevant section.
- Add new line items as needed.
- Entries should follow the [Common Changelog Style Guide](https://github.com/vweevers/common-changelog). 
- Remove all unused sections before merging.
- Proof-read.
-->

### Added
- Added a WordPress version compatibility check before loading the remote data blocks integration.

## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [ ] This change works and has been tested on a sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test

1. Check out PR.
2. Setup local dev environment with MU Plugins pointed to this checked out branch.

	```sh
	# Create a dev environment from a VIP deployed app with integrations configured and VIP GO MU Plugins pointed to the checked out branch
	vip dev-env create @app_id.environment_type -s <local_dev_environment_slug> -u <path_to_checked_out_branch>

	# Start the dev environment
	vip dev-env start <local_dev_environment_slug> --vscode
	```

3. When the VS Code window opens up, edit the `integrations/integrations-config/integrations.json` file to add a `remote-data-blocks` integration config -

	```json
	"remote-data-blocks": {
		"type": "remote-data-blocks",
		"org": {
			"status": "enabled"
		},
		"env": {
			"status": "enabled",
			"config": {
				"version": "latest",
				"minimum_wp_version": "6.7"
			}
		}
	}
	```

4. Verify that RDB integration is loaded as default WP version is greater than 6.7. 
5. Try changing the local dev environment's WordPress version to to different version and verify that RDB integration is loaded or not loaded based on the WordPress version.

	```sh
	# Change the WordPress version of the local dev environment
	vip dev-env update -s <local_dev_environment_slug> -w <wordpress_version>

	# Restart the local dev environment
	vip dev-env start <local_dev_environment_slug>
	```

shekharnwagh committed:
feat: skip loading remote data blocks on unsupported WP versions

Adds version compatibility check before loading the remote data blocks integration.
The integration will only load if the current WordPress version meets or exceeds
the minimum_wp_version specified in config. If no minimum version is configured,
the integration loads normally.

shekharnwagh committed:
try: avoid global in rdb tests

ingeniumed commented on integrations/remote-data-blocks.php:
```diff
@@ -25,6 +25,33 @@ public function is_loaded(): bool {
+	 *              WordPress version defined in the environment config or if no minimum WordPress
+	 *              version is defined, `false` otherwise.
+	 */
+	public function is_supported_wp_version(): bool {
```
I would move this to the base `integration` class instead, because this isn't going to be the only integration that would use it. That way, every integration only needs line 73 to actually prevent loading the integration.

ingeniumed replied:
I understand why this is done this way, ignore this for now.

ingeniumed commented on integrations/remote-data-blocks.php:
```diff
@@ -25,6 +25,33 @@ public function is_loaded(): bool {
 		return defined( 'REMOTE_DATA_BLOCKS__LOADED' );
 	}
 
+	public function get_wp_version(): string {
```
It would be good to add a comment here as to why this is implemented this way.

shekharnwagh replied:
a12fdbf53

shekharnwagh committed:
add comments to explain wrapping global wp_version var

chriszarate commented on integrations/remote-data-blocks.php:
```diff
@@ -25,6 +25,38 @@ public function is_loaded(): bool {
+	 * Wraps the global `$wp_version` variable in a function to make it easier to mock in tests.
+	 */
+	public function get_wp_version(): string {
+		global $wp_version;
```
In this codebase, I more commonly see `get_bloginfo( 'version' )` to obtain the current WP version.

shekharnwagh replied:
I found `global $wp_version` being used in 38 files vs `get_bloginfo( 'version' )` used in 15 files, if that matters. TBH I only searched for the first and it, being so widely used, settled on that without checking if there is any other way. The `get_bloginfo( 'version' )` does seem better, given that it's a function which might be easier to mock.

chriszarate commented on integrations/remote-data-blocks.php:
```diff
@@ -25,6 +25,38 @@ public function is_loaded(): bool {
+
+		$config = $this->get_env_config();
+		if (
+			isset( $config['minimum_wp_version'] ) &&
+			is_string( $config['minimum_wp_version'] ) &&
```
Shouldn't this method return `false` if `$config['minimum_wp_version']` is not correctly set, instead of defaulting to `true`?

shekharnwagh replied:
Made it return `true` to account for case where GOOP change was deployed after this or the configmap changes weren't propagated. Now that GOOP is deployed, I think we could default to being strict where if `minimum_wp_version` is not in config we don't load the plugin. 

Though `minimum_wp_version` acts like an additional filter, and wouldn't absence of a filter indicate that needn't be applied ?

maxschmeling replied:
> Though minimum_wp_version acts like an additional filter, and wouldn't absence of a filter indicate that needn't be applied ?

This is how I would look at it. If the value isn't there, don't apply the logic and let it load like normal.

chriszarate replied:
I see that logic. But since this is designed to prevent fatals on customer sites, I was thinking more from a security / protective stance. If we expect this guard to always be in place, do we want it to "fail open" (permissive) or "fail closed" (restrictive)?

chriszarate replied:
Ie, what happens if a bug in the code that provides this config accidentally stops providing `minimum_wp_version`? What do we want to happen?

shekharnwagh replied:
Moved to skip loading RDB when `minimum_wp_version` is not avaiable

chriszarate commented on integrations/remote-data-blocks.php:
```diff
@@ -43,6 +75,11 @@ public function load(): void {
 			}
 
+			if ( ! $this->is_supported_wp_version() ) {
+				$this->is_active = false;
```
What is the point of `$this->is_active`? It is never read, nor is it declared as a public class var.

shekharnwagh replied:
It is exposed via this public method in the parent - https://github.com/Automattic/vip-go-mu-plugins/blob/ddc0ff307245ae183e9a56f984115fba3d49bf3a/integrations/integration.php#L93-L95

chriszarate commented on tests/integrations/test-remote-data-blocks.php:
```diff
@@ -215,4 +215,67 @@ public function test_get_versions_ignores_non_matching_directories(): void {
+			'no minimum_wp_version in config' => [ '6.7', [], true ],
+			'wp_version less than minimum'    => [ '5.6', [ 'minimum_wp_version' => '6.7' ], false ],
+			'wp_version equal to minimum'     => [ '6.7.1', [ 'minimum_wp_version' => '6.7' ], true ],
+			'wp_version greater than minimum' => [ '6.8.2', [ 'minimum_wp_version' => '6.7' ], true ],
```
Could use some tests with patch versions and equalities
```suggestion
			'wp_version greater than minimum' => [ '6.8.2', [ 'minimum_wp_version' => '6.7' ], true ],
			'wp_version equal to minimum' => [ '6.7', [ 'minimum_wp_version' => '6.7' ], true ],
			'wp_version equal to minimum but more specific' => [ '6.7.0', [ 'minimum_wp_version' => '6.7' ], true ],
			'wp_version equal to minimum but less specific' => [ '6.7', [ 'minimum_wp_version' => '6.7.0' ], true ],
```

shekharnwagh replied:
> 'wp_version equal to minimum but less specific' => [ '6.7', [ 'minimum_wp_version' => '6.7.0' ], true ]

This fails since `version_compare` treats `6.7` as less than `6.7.0`. Is this something which might happen with wordpress versions ? If yes, we could patch the logic to handle such cases. When I check the WP releases page - https://wordpress.org/download/releases/, the versions were like `6.7`, `6.7.1`, `6.7.2` with `6.7.0` not mentioned. That might indicate we don't get cases like `6.7.0` but I don't see anything wrong with that. If we decide to patch it should we only worry about patch version being absent/present or extend that for other cases where versions contains texts like `release`, `rc`, `beta` ?

chriszarate replied:
I don't know the answer, kinda why I wanted more test cases. I bet @rinatkhaziev has good advice here

rinatkhaziev replied:
WP doesn't follow semver. So it's '6.7' => '6.7.1'. pre-releases are `X.Y-beta1', 'X.Y-rc1`, etc.

shekharnwagh replied:
Added few more tests for pre-release versions, but the way it is now, works for above scenarios

chriszarate commented on tests/integrations/test-remote-data-blocks.php:
```diff
@@ -215,4 +215,67 @@ public function test_get_versions_ignores_non_matching_directories(): void {
+		return [
+			'no minimum_wp_version in config' => [ '6.7', [], true ],
+			'wp_version less than minimum'    => [ '5.6', [ 'minimum_wp_version' => '6.7' ], false ],
+			'wp_version equal to minimum'     => [ '6.7.1', [ 'minimum_wp_version' => '6.7' ], true ],
```
they are not equal

shekharnwagh committed:
add more wp version compare tests

Co-authored-by: Chris Zarate <chris.zarate@automattic.com>

shekharnwagh committed:
remove wp version compare test for less specific wp version than minimum_wp_version

shekharnwagh committed:
skip load if minimum_wp_version unspecified

maxschmeling reviewed (approved):
I'm marking this as approved since I tested it locally and it's working. Still need to address open discussion items.

shekharnwagh commented:
> I'm marking this as approved since I tested it locally and it's working. Still need to address open discussion items.

I think I've replied to or addressed all the comments. Waiting for review and an approval from someone from Cantina.
